tests/BookReader.test.js

```diff
@@ -1,59 +1,154 @@
+    // Call it immediately
+    debouncedFunc();
+    expect(func).toHaveBeenCalledTimes(0); // func not called
+
+    // Call it several times with 500ms between each call
+    for (let i = 0; i < 10; i++) {
+      clock.tick(500);
+      debouncedFunc();
+    }
+    expect(func).toHaveBeenCalledTimes(0); // func not called
+
+    // wait 1000ms
+    clock.tick(1000);
+    expect(func).toHaveBeenCalledTimes(1); // func called
+  });
 });
 
-afterEach(() => {
-  clock.restore();
+describe('Testing Throttle function', () => {
+  let clock;
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers();
+  });
+  afterEach(() => {
+    clock.restore();
+  });
+
+  test('testing throttle', () => {
+    const callMethod = jest.fn();
+    const throttled = BookReader.util.throttle(() => callMethod(), 100);
+
+    const t = setInterval(() => {
+       throttled()
+    }, 50);
+
+    clock.tick(400);
+    expect(callMethod).toHaveBeenCalledTimes(4);
+    clearInterval(t);
+  });
 });
 
-test('testing debounce', () => {
-  const func = jest.fn();
-  const debouncedFunc = BookReader.util.debounce(func, 1000);
-  // Call it immediately
-  debouncedFunc();
-  expect(func).toHaveBeenCalledTimes(0); // func not called
+describe("Trigger function test", () => {
+  beforeEach(() => {
+    BookReader.prototype.trigger('SKR', 'val');
+  });
+  test("testing trigger", () => {
+    $(document).trigger('BookReader:' + 'SKR', 'val');
+  });
+});
 
-  // Call it several times with 500ms between each call
-  for(let i = 0; i < 10; i++) {
-    clock.tick(500);
-    debouncedFunc();
-  }
-  expect(func).toHaveBeenCalledTimes(0); // func not called
+describe("Testing extendParams", () => {
+  const obj1 = {
+    prop1: "SK",
+    prop2: 20
+  };
+  const obj2 = {
+    prop2: 19,
+    prop3: "SKR"
+  };
+  var modifiedNewParams = $.extend({}, obj2);
+  beforeEach(() => {
+    BookReader.prototype.extendParams(obj1, obj2);
+  });
+  test("jquery.extend", () => {
+    $.extend(obj1, modifiedNewParams);
+  });
+});
+
+describe("Testing jquery bind", () => {
+  function you() {
+    alert( "User clicked on 'you.'" );
+  };
+  beforeEach(() => {
+    BookReader.prototype.bind('SK', you);
+  });
+  test("jquery.bind", () => {
+    $(document).bind('BookReader:' + 'SK', you);
+  });
+});
 
-  // wait 1000ms
-  clock.tick(1000);
-  expect(func).toHaveBeenCalledTimes(1);  // func called
+describe("Testing unbind function", () => {
+  function fox() {
+    alert( "The quick brown fox jumps over the lazy dog." );
+  };
+  beforeEach(() => {
+    BookReader.prototype.unbind('SKR', fox);
+  });
+  test("jquery unbind", () => {
+    $(document).bind('BookReader:' + 'SKR', fox);
+  });
 });
 
+describe("Testing setUpKeyListeners function", () => {
+  function key() {
+    alert( "Handler for .keydown() called." );
+  };
+  beforeEach(() => {
+    BookReader.prototype.setupKeyListeners();
+  });
+  test("jquery keydown", () => {
+    $(document).keydown(key);
+  });
```

Comment:
I think for this test, you want to check to see if something has happened after the keypress has occurred. For instance, if you look inside the `setupKeyListeners` method, you can see when the user presses the left or right keys, we call `self.left()` or `self.right()`.

What I would do here is create a book reader object, override the `left` and `right` methods to do something that you can check against. ie:

```
var br = new BookReader();
br.left = jest.fn();  // mock out the `left` function
br.setupKeyListeners(); // set up the book reader keylisteners
$(document).keydown(key); // I'm not exactly sure how to simulate the "left" key getting pressed
expect(br.left).toHaveBeenCalledTimes(1); // check to make sure `left` got called
```
---
Also since you only have a single test in here, you don't necessarily need `beforeEach`. It doesn't hurt, but it's more useful if you have a bunch of tests that need the same setup and since you just have a single test, it doesn't really help that much.

Reply:
noted, I will not add beforeEach and afterEach for single tests